content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+---
+title: "RFC: Extension Packaging & Lookup"
+slug: rfc-extension-packaging-lookup
+date: 2024-11-04T19:07:44Z
+lastMod: 2024-11-04T19:07:44Z
+description: |
+  A proposal to modify the PostgreSQL core so that all files required for an
+  extension live in a directory named for the extension, along with a search
+  path to find extension directories.
+tags: [Postgres, Extensions, RFC, Packaging, Kubernetes, Docker, Packaging, Postgres.app]
+type: post
+draft: true
+---
+
+A few weeks ago, I started [a pgsql-hackers thread] proposing a new extension
+file organization and a search path [GUC] for finding extensions. The
+[discussion] of [Christoph Berg]'s [`extension_destdir` patch][destdir]
+inspired this proposal. These threads cover quite a lot of territory, so I
+thought it would be useful to pull together a more unified, public proposal.
+
+## The Problem
+
+A number of challenges face extension users in various configurations, thanks
+to extension file organization in the Postgres core. The common thread among
+them is the need to add extensions without changing the Postgres installation
+itself.
+
+### Packager Testing
+
+On Debian systems, the user account that creates extension packages lacks
+permission to add files to Postgres install. But testing extensions requires
+installing the extension files where Postgres can find them. Furthermore,
+extensions should ideally be built against a clean Postgres install; adding an
+extension in order to run `make installcheck` would pollute it.
+
+[Christoph's patch][destdir] solves these problems by adding a second lookup
+path for extensions and dynamic modules, so that Postgres can load them
+directly from the package build directory.
+
+Alas, the patch isn't ideal, because it simply specifies a prefix and appends
+the full `pg_config` directory paths to it. For example, if `--sharedir`
+outputs `/opt/share` and `extension_destdir` GUC is set to `/tmp/build/myext`,
+the patch will search in `/tmp/build/myext/opt/share`. This approach works for
+the packaging use case, which explicitly uses full paths with a prefix, but
+would be weird for other use cases.
+
+### Docker Immutability
+
+Docker images are immutable. To install persistent extensions in a running
```

Comment:
Container images are immutable. The filesystem where a container image runs has a writeable top layer, but is not persistent. So maybe a precision can be made here about this.

Reply:
Use "container images" or "OCI images" instead of "Docker images".

Reply:
Done in 15c4952.